src/libproc_macro/bridge.rs

```diff
@@ -0,0 +1,856 @@
+                })*
+            })*
+        }
+    }
+    with_api!(Self, self_, define_handlify_impls);
+
+    impl Expand1 {
+        pub fn run<S: Server>(
+            &self,
+            server: S,
+            input: S::TokenStream,
+        ) -> thread::Result<S::TokenStream> {
+            let server = Handlify::new(self.counters, server);
+            let input = super::TokenStream::from_data(input, &mut server.storage.borrow_mut());
+            let output = (self.run)(&(&server as &super::Server), input);
+            output.map(|output| output.to_data(&mut server.storage.borrow_mut()))
+        }
+    }
+
+    impl Expand2 {
+        pub fn run<S: Server>(
+            &self,
+            server: S,
+            input: S::TokenStream,
+            input2: S::TokenStream
+        ) -> thread::Result<S::TokenStream> {
+            let server = Handlify::new(self.counters, server);
+            let input = super::TokenStream::from_data(input, &mut server.storage.borrow_mut());
+            let input2 = super::TokenStream::from_data(input2, &mut server.storage.borrow_mut());
+            let output = (self.run)(&(&server as &super::Server), input, input2);
+            output.map(|output| output.to_data(&mut server.storage.borrow_mut()))
+        }
+    }
+
+    pub trait Registry {
+        fn register_custom_derive(
+            &mut self,
+            trait_name: &str,
+            expand: Expand1,
+            attributes: &[&'static str],
+        );
+
+        fn register_attr_proc_macro(
+            &mut self,
+            name: &str,
+            expand: Expand2,
+        );
+
+        fn register_bang_proc_macro(
+            &mut self,
+            name: &str,
+            expand: Expand1,
+        );
+    }
+}
+
+trait HandleFromData<S: server::Types> {
+    type Data;
+    fn from_data(x: Self::Data, storage: &mut Storage<S>) -> Self;
+}
+
+trait HandleToData<'s, S: server::Types> {
+    type Data;
+    fn to_data(self, storage: &'s mut Storage<S>) -> Self::Data;
+}
+
+// FIXME(eddyb) achieve ABI compatibility for these types.
+impl<S: server::Types, H: HandleFromData<S>> HandleFromData<S> for Option<H> {
+    type Data = Option<H::Data>;
+    fn from_data(x: Self::Data, storage: &mut Storage<S>) -> Self {
+        x.map(|x| H::from_data(x, storage))
+    }
+}
+
+macro_rules! passthrough_from_and_to_data {
+    ($($ty:ty),*) => {
+        $(
+            impl<S: server::Types> HandleFromData<S> for $ty {
+                type Data = Self;
+                fn from_data(x: Self::Data, _: &mut Storage<S>) -> Self { x }
+            }
+            impl<S: server::Types> HandleToData<'_, S> for $ty {
+                type Data = $ty;
+                fn to_data(self, _: &mut Storage<S>) -> Self::Data { self }
+            }
+        )*
+    }
+}
+passthrough_from_and_to_data! {
+    (),
+    bool,
+    char,
+    &'a [u8],
+    &'a str,
+
+    // FIXME(eddyb) achieve ABI compatibility for these types.
+    String,
+    PathBuf,
```

Comment:
Could this show an example of how `PathBuf` for example would be ABI-safe to cross the boundary? (just so we've got the code written down in case we need to do it for other types).

Reply:
I'll try to implement https://github.com/rust-lang/rust/pull/49219#issuecomment-385094778 before merging, which would alleviate this.